### DISCUSSION TITLE
https://thorsten.suckow-homberg.de/blog/2023/04/14/big-ball-of-mud-decay

### DISCUSSION THREAD

**Discussion by giscus** [ThorstenSuckow/thorsten.suckow-homberg.de#78] [Blog]
# https://thorsten.suckow-homberg.de/blog/2023/04/14/big-ball-of-mud-decay

Ein Kommentar zu einem Kommentar zu Eberhard Wolff's Episode 159 - Big Ball of Mud als Teil von Software-Architektur im Stream. Eine englische Übersetzung findet sich hier.

https://thorsten.suckow-homberg.de/blog/2023/04/14/big-ball-of-mud-decay

**Comment by ewolff** [comment]
Danke für den ausführlichen Blog-Post - mir ist nun deutlich klarer, was Du im Stream gemeint hast! Natürlich finde ich Tests sinnvoll wie auch eine saubere Struktur eines System eigentlich sinnvoll ist. Allerdings muss eine saubere Struktur auch einen Geschäftswert haben. Da ich sehr viele Projekte sehe, die genau so eine saubere Struktur etablieren wollen, um die Wartbarkeit zu verbessern, ist sie in vielen Fällen ganz sicher vorhanden und auch erkannt. Ich finde die Broken-Windows-Idee allerdings schwierig: In der Praxis sehe ich praktisch nie perfekt strukturierte Systeme. Also gibt es immer ein paar "broken Windows". Die alle zu "reparieren" ist meiner Meinung nach nicht nur sehr aufwändig, sondern auch unwirtschaftlich. Wenn ein Systemteil nie oder nur selten geändert wird, hat die saubere Struktutrierung keine positiven wirtschaftlichen Auswirkungen. Warum sollte man dann diesen Systemteil verbessern? 

Aus dem Blog könnte man den Eindruck gewinnen, dass man eine saubere Strukturierung haben will, damit man Tests schreibt. Das wäre für mich kein ausreichender Grund, um die Strukturierung aufzuräumen, weil es ein indirekter Effekt der besseren Strukturierung ist. Das kann als Begründgung  für teuere Verbesserungen eigentlich nicht ausreichen. Und wer sich durch schlechte Struktur davon abhalten lässt, Tests zu schreiben, kann niemals die schlechte Struktur mit einem Sicherheitsnetz aus Tests verbessern.

Übrigens stellt Evan Big Ball of Mud gerade ähnlich wie die Aussage von Hunt und Thomas als eine Möglichkeit dar, um einen Bereich des Systems mit schlechtzer Strukturierung abzugrenzen und die weitere Ausbreitung des Problems zu vermeiden, oder eben das "Sweeping it under the rug". Von daher sehe ich da keinen Widerspruch zu Big Ball of Mud und den anderen Pattern. 

Ein Hinweis noch: Ich finde es schwierig, wenn Manager:innen Tests unterbinden. Dann ist die funktionale Korrektheit des Systems gefährdet und das kann eigentlich nicht ernsthaft zu einem akzeptablen Projekt-Ergebnis führen. Personen, die das dennoch glauben, sollten vielleicht keine Software-Projekte leiten. Zumindest sollte es im Team Personen geben, die dieses fundamentale Prinzip kennen. Warum hören Manager:innen dann auf sie nicht? Am Ende ist das Test-Verbot dann eher ein soziales Problem. Außerdem frage ich mich, wie die Manager:in merkt, ob Entwickler:innen Unit Tests schreiben oder nicht. Schauen die ernsthaft auf der Ebene in den Code? Können Entwickler:innen dann nicht einfach trotzdem Tests schreiben?

**Reply by ThorstenSuckow** [reply]
Ich gebe Dir recht, die Broken Windows sind immer da (aus Erfahrung ist eine Unterscheidung zwischen Sprung im und zerborstenem Glas manchmal nötig). Ab wann sind solche Broken Windows aber nicht mehr wesentlich für die Beurteilung eines Systems, also ob es (auch auf längere Sicht) wirtschaftlich Erfolg bringt und Sinn macht, und ob die Entwickler in der Arbeit daran einen Sinn sehen, so daß die Entwicklung intrinsisch erfolgt, und keinem bloßen wirtschaftlichen Interesse dient. (Ich würde behaupten: Wenn ein BBOM kein wirtschaftliches Problem ist, dann kann er aber zu einem sozialen Problem werden.)

Das Unternehmen sagt: Ein sauber strukturiertes System hat keine positiven wirtschaftlichen Auswirkung gegenüber dem derzeitigen Stand, also strukturieren wir nicht sauber. Der Entwickler, der sich innerhalb des BBOM bewegen muß, kann vielleicht keine ganz so sachliche Aussage treffen, weil er emotional verbunden ist mit der Entwicklung und der Aufrechterhaltung des Systems. Stösst er bei der Implementierung von Tests, die für den BBOM geschrieben werden, auf entsprechende Hürden (bspw. stark gekoppelte Programme, die sich nur schwierig mocken lassen; Funktionen, deren Seiteneffekte ein ganzes Programm in einen anderen Zustand versetzen; Whitebox-Tests, die die Implementierung einer Funktion nachstellen, weil die eigentlich zu testende Funktion schwer zu verstehen ist), dann wird wahrscheinlich das Testen dieses BBOM nicht in dem Maße von Quantität und Qualität erfolgen, wie es eigentlich notwendig wäre (die Notwendigkeit vorausgesetzt). 

Gräser, Bäume, Ställe und Hütten nimmt man aus der obersten Etage eines Elfenbeinturms gerade noch so wahr, wenn überhaupt. Den Umstand vorausgesetzt, wichtige Personen eines Unternehmens sitzen genau dort, wird es sich im übertragenen Sinne ähnlich mit der Beurteilung des Gesamtsystems verhalten: Fern von Zuweisungen, Kontrollstrukturen, Funktionen, Klassen, Modul- und fachlichen Grenzen und der Infrastruktur - der BBOM fällt dann nicht ins Gewicht, wenn das System im Wortsinne produktiv ist. Dagegen findet sich ein Entwickler - außerhalb dieses Turms - bei dem Versuch, den BBOM zu verstehen, vielleicht eher dabei wieder, in eine Mandelbrotmenge hineinzustürzen. (Mancher wird vielleicht auch lieber Joseph Conrad zitieren wollen.)

Technische Schulden, die durch so einen BBOM entstehen, bekommt man dann ja auch zu spüren, wenn neue Entwickler eingearbeitet werden müssen. Ich gehe aber davon aus, dass Unternehmen das bei der Beurteilung von wirtschaftlichem Wert berücksichtigen.  

Wie weit läßt man den BBOM dann wachsen? "Sweeping it under the rug" als Demarkierung und damit (hoffentlich) einhergehendes Information Hiding hilft dabei (gglf. mit Hilfe von Anticorruption Layer), die Schnittstellen zu dem BBOM zu realisieren. Ab dem Punkt kann dann vielleicht auch ein Entwickler das System aus der Distanz - in einem neuen Bezugssystem - anders betrachten. 

------------
Edit:
In diesem Kommentar wurde nachträglich das Unternehmen aus dem Weltall zurück auf die Erde in den Elfenbeinturm gesetzt, um eine eher bekannte Metapher zu nutzen: In der Software Entwicklung bewohnt oft ein nicht-codender Architekt diesen Turm. Ich habe mir erlaubt, für das vorliegende Beispiel den (nicht unbedingt fehlenden) Architekten durch einen Teil des Management auszutauschen.
